### ISSUE TITLE
[no-unused-vars] False positive with Decorators and Promises

### ISSUE THREAD

**Issue by msanguineti** [typescript-eslint/typescript-eslint#2969]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```sh
# reproduction repo
git clone https://github.com/msanguineti/ts-eslint-test.git
# or
npx degit https://github.com/msanguineti/ts-eslint-test.git#main
```

`.eslintrc.json`
```JSON
{
  "env": {
    "es2021": true,
    "node": true
  },
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking"
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 12,
    "sourceType": "module",
    "project": "tsconfig.json"
  },
  "plugins": ["@typescript-eslint"],
  "rules": {}
}
```

`index.ts`
```TS
interface Message {
  body: string
}

const LogParameter = (): ParameterDecorator => {
  return (): void => {
    // ...
  }
}

export class Messenger {
  async deliverMessage(@LogParameter() message: string): Promise<Message> {
    return Promise.resolve({ body: message })
  }
}
```

`tsconfig.json`
```JSON
{
  "compilerOptions": {
    "module": "commonjs",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "ES2018",
    "strict": true
  },
  "exclude": ["node_modules", "dist"]
}
```

**Expected Result**

No linting errors or warnings

**Actual Result**

```sh
  1:11  warning  'Message' is defined but never used  @typescript-eslint/no-unused-vars
```

**Additional Info**

This problem did not happen with `@typescript-eslint/parser` version `4.13.0`. It only appeared in the last few days when I've upgraded the package to the latest version

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.14.0` |
| `@typescript-eslint/eslint-plugin`| `4.14.0`|
| `TypeScript`                | `4.1.3` |
| `ESLint`                    | `7.18.0` |
| `node`                      | `14.15.4` |

**Comment by bradzacher** [comment]
Same as #2941, #2942, #2945, #2947, #2950, #2957

Please use the search next time.
